### PR TITLE
Test stability

### PR DESCRIPTION
**Pull request by jglick**
- [X] Rerun flaky tests
- [X] Fix some code bugs thought to cause random test failures
- [X] Use more reliable test idioms ([JENKINS-25975](https://issues.jenkins-ci.org/browse/JENKINS-25975))

@reviewbybees

### CODE REVIEW

**Comment by jglick** [comment]
Hmm, seems the CI builder passes `-Dignore.random.failures` on its own, so I really need to remove all `@RandomlyFails`.

**Comment by jglick** [comment]
`buildShellScriptAcrossDisconnect` flaked; `executorStepRestart` failed repeatedly.

**Comment by jglick** [comment]
`executorStepRestart` flaked again.

**Comment by jglick** [comment]
fcff78c ran without any flakes.

**Comment by jglick** [comment]
This time `buildShellScriptAcrossDisconnect` failed three times in a row.

**Comment by tfennelly** [comment]
:+1: nice

**Comment by stephenc** [comment]
:+1: from the blind man on the galloping horse

**Comment by jtnord** [comment]
:+1:

File: cps/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsStepContext.java [hunk 1]
```diff
@@ -303,19 +306,24 @@ public CpsBodyInvoker newBodyInvoker(BodyReference body) {
...
-        if (t==null)
+    @Override public synchronized void onFailure(Throwable t) {
+        if (t == null) {
             throw new IllegalArgumentException();
-        if (isCompleted())
-            throw new IllegalStateException("Already completed", t);
+        }
+        if (isCompleted()) {
+            LOGGER.log(Level.WARNING, "already completed " + this, new IllegalStateException(t));
+            return;
```
Comment (jglick) [inline cps/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsStepContext.java]: [JENKINS-27554](https://issues.jenkins-ci.org/browse/JENKINS-27554)

**Pull request merged**